### PR TITLE
Refactor Volumes to use InternalVolume instead of API types

### PR DESCRIPTION
**Pull request by markturansky**
@thockin Per our discussion in #6695.

I refactored the API volume types out of the volume plugin interfaces and replaced it with InternalVolume.  We agreed on an internal "volume source" but nearly all plugins also used the volume's name, so an InternalVolume construct made more sense.

### CODE REVIEW

**Comment by dchen1107** [comment]
cc/ @vishh

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 1]
```diff
@@ -71,19 +69,27 @@ func (plugin *awsElasticBlockStorePlugin) GetAccessModes() []api.AccessModeType
...
+func (plugin *awsElasticBlockStorePlugin) NewBuilder(spec *volume.InternalVolume, podRef *api.ObjectReference, _ volume.VolumeOptions) (volume.Builder, error) {
 	// Inject real implementations here, test through the internal function.
 	return plugin.newBuilderInternal(spec, podRef.UID, &AWSDiskUtil{}, mount.New())
 }
 
-func (plugin *awsElasticBlockStorePlugin) newBuilderInternal(spec *api.Volume, podUID types.UID, manager pdManager, mounter mount.Interface) (volume.Builder, error) {
-	volumeID := spec.AWSElasticBlockStore.VolumeID
-	fsType := spec.AWSElasticBlockStore.FSType
+func (plugin *awsElasticBlockStorePlugin) newBuilderInternal(spec *volume.InternalVolume, podUID types.UID, manager pdManager, mounter mount.Interface) (volume.Builder, error) {
+
```
Comment (thockin) [inline pkg/volume/aws_ebs/aws_ebs.go]: blank line

File: pkg/volume/empty_dir/empty_dir.go [hunk 2]
```diff
@@ -66,27 +66,29 @@ func (plugin *emptyDirPlugin) Name() string {
...
 
-func (plugin *emptyDirPlugin) CanSupport(spec *api.Volume) bool {
+func (plugin *emptyDirPlugin) CanSupport(spec *volume.InternalVolume) bool {
 	if plugin.legacyMode {
 		// Legacy mode instances can be cleaned up but not created anew.
 		return false
 	}
 
-	if util.AllPtrFieldsNil(&spec.VolumeSource) {
+	if util.AllPtrFieldsNil(spec.PersistentVolumeSource) ||
```
Comment (thockin) [inline pkg/volume/empty_dir/empty_dir.go]: I think you want && here

File: pkg/volume/empty_dir/empty_dir.go [hunk 3]
```diff
@@ -66,27 +66,29 @@ func (plugin *emptyDirPlugin) Name() string {
...
+		util.AllPtrFieldsNil(spec.VolumeSource) {
```
Comment (thockin) [inline pkg/volume/empty_dir/empty_dir.go]: no need to linewrap this - less readable

File: pkg/volume/empty_dir/empty_dir.go (hunk 2 above)
Comment (thockin) [inline pkg/volume/empty_dir/empty_dir.go]: This is a bit weird, actually.  It means that EmptyDir will trigger in the case of a malformed PVS, which is NOT what we want.

Perhaps we should make the internal struct hold pointers to PVS and VS, and that way EmptyDir would only consider if spec.VS != nil && AllPtrFieldsNil(spec.VS)

What think?

File: pkg/volume/git_repo/git_repo.go [hunk 4]
```diff
@@ -119,9 +116,11 @@ func (gr *gitRepo) SetUp() error {
 }
 
 // This is the spec for the volume that this plugin wraps.
-var wrappedVolumeSpec = &api.Volume{
-	Name:         "not-used",
-	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{}},
+var wrappedVolumeSpec = &volume.InternalVolume{
+	Name: "not-used",
```
Comment (thockin) [inline pkg/volume/git_repo/git_repo.go]: did gofmt unalign these?  I am surprised - please run gofmt just to be sure.

File: pkg/volume/host_path/host_path.go [hunk 5]
```diff
@@ -47,11 +47,8 @@ func (plugin *hostPathPlugin) Name() string {
 	return hostPathPluginName
 }
 
-func (plugin *hostPathPlugin) CanSupport(spec *api.Volume) bool {
-	if spec.HostPath != nil {
-		return true
-	}
-	return false
+func (plugin *hostPathPlugin) CanSupport(spec *volume.InternalVolume) bool {
+	return spec.VolumeSource.HostPath != nil || spec.PersistentVolumeSource.HostPath != nil
```
Comment (thockin) [inline pkg/volume/host_path/host_path.go]: Do we still need PVS.HostPath?  I thought that was no longer needed?

File: pkg/volume/iscsi/iscsi.go [hunk 6]
```diff
@@ -52,8 +52,8 @@ func (plugin *ISCSIPlugin) Name() string {
 	return ISCSIPluginName
 }
 
-func (plugin *ISCSIPlugin) CanSupport(spec *api.Volume) bool {
-	if spec.ISCSI == nil {
+func (plugin *ISCSIPlugin) CanSupport(spec *volume.InternalVolume) bool {
+	if spec.VolumeSource.ISCSI == nil {
```
Comment (thockin) [inline pkg/volume/iscsi/iscsi.go]: no PVS for iSCSI?

File: pkg/volume/nfs/nfs.go [hunk 7]
```diff
@@ -50,11 +50,8 @@ func (plugin *nfsPlugin) Name() string {
 	return nfsPluginName
 }
 
-func (plugin *nfsPlugin) CanSupport(spec *api.Volume) bool {
-	if spec.VolumeSource.NFS != nil {
-		return true
-	}
-	return false
+func (plugin *nfsPlugin) CanSupport(spec *volume.InternalVolume) bool {
+	return spec.VolumeSource.NFS != nil
```
Comment (thockin) [inline pkg/volume/nfs/nfs.go]: no PVS for NFS?

File: pkg/volume/plugins.go [hunk 8]
```diff
@@ -119,6 +119,39 @@ type VolumePluginMgr struct {
 	plugins map[string]VolumePlugin
 }
 
+// InternalVolume is an internal representation of a volume.  All API volume types translate to InternalVolume.
+type InternalVolume struct {
```
Comment (thockin) [inline pkg/volume/plugins.go]: suggest `type Spec`

Any particular reason to break out VolumeSource?  I would just list them directly here.  It's simpler and it is self-contained.

```
type Spec struct {
    Name string
    // Exactly one of the following must be set.
    VolumeSource *api.VolumeSource
    PersistentVolumeSource *api.PersistentVolumeSource
}
```

File: pkg/volume/plugins.go [hunk 9]
```diff
@@ -119,6 +119,39 @@ type VolumePluginMgr struct {
...
+}
+
+// InternalVolumeSource is a union structure encompassing all types of volumes.
+// This union allows us to separate API types but treat all volumes equally as plugins.
+type InternalVolumeSource struct {
+	api.VolumeSource
+	api.PersistentVolumeSource
+}
+
+func ConvertFromVolume(vs *api.Volume) *InternalVolume {
```
Comment (thockin) [inline pkg/volume/plugins.go]: NewSpecFromVolume

File: pkg/volume/plugins.go [hunk 10]
```diff
@@ -119,6 +119,39 @@ type VolumePluginMgr struct {
...
+	return &InternalVolume{
+		Name: vs.Name,
+		InternalVolumeSource: InternalVolumeSource{
+			vs.VolumeSource,
+			api.PersistentVolumeSource{},
+		},
+	}
+}
+
+func ConvertFromPersistentVolume(pv *api.PersistentVolume) *InternalVolume {
```
Comment (thockin) [inline pkg/volume/plugins.go]: NewSpecFromPersistentVolume

File: pkg/volume/plugins_test.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+package volume
+
+import (
+	"testing"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+)
+
+func TestInternalVolumeSourceConverters(t *testing.T) {
+
```
Comment (thockin) [inline pkg/volume/plugins_test.go]: blank line - you love these for some reason I don't understand :)

File: pkg/volume/plugins.go [hunk 12]
```diff
@@ -119,6 +119,39 @@ type VolumePluginMgr struct {
...
+	api.VolumeSource
+	api.PersistentVolumeSource
+}
+
+func ConvertFromVolume(vs *api.Volume) *InternalVolume {
+	return &InternalVolume{
+		Name: vs.Name,
+		InternalVolumeSource: InternalVolumeSource{
+			vs.VolumeSource,
+			api.PersistentVolumeSource{},
```
Comment (thockin) [inline pkg/volume/plugins.go]: don't init the field you don't need - it's just noise.  Use a named init here and it will be clearer

File: pkg/volume/empty_dir/empty_dir.go (hunk 2 above)
Comment (markturansky) [inline pkg/volume/empty_dir/empty_dir.go]: EmptyDir has the same issue with a malformed VS (though by this point validation should have caught it).

I think && would preserve the current behavior of a default EmptyDir, if that's what we want.  The || is definitely wrong.

File: pkg/volume/git_repo/git_repo.go (hunk 4 above)
Comment (markturansky) [inline pkg/volume/git_repo/git_repo.go]: I always gofmt because I've gotten bitten enough by hack/verify-gofmt.sh.

File: pkg/volume/host_path/host_path.go (hunk 5 above)
Comment (markturansky) [inline pkg/volume/host_path/host_path.go]: It's what I use for development and testing.  The API comments explicitly flag it as dev/test use only.

File: pkg/volume/iscsi/iscsi.go (hunk 6 above)
Comment (markturansky) [inline pkg/volume/iscsi/iscsi.go]: It is not in the PersistentVolumeSource struct.  I figured adding it would be a different PR, as you've suggested in other PRs in the recent past (the NFS pointer made its way into one).

File: pkg/volume/nfs/nfs.go (hunk 7 above)
Comment (markturansky) [inline pkg/volume/nfs/nfs.go]: Same as iSCSI.  Not currently in the PSV struct.

File: pkg/volume/plugins.go (hunk 8 above)
Comment (markturansky) [inline pkg/volume/plugins.go]: I like this better.  Will refactor to it.

File: pkg/volume/plugins_test.go (hunk 11 above)
Comment (markturansky) [inline pkg/volume/plugins_test.go]: You hate them for reasons I don't understand!

File: pkg/volume/empty_dir/empty_dir.go (hunk 2 above)
Comment (thockin) [inline pkg/volume/empty_dir/empty_dir.go]: EmptyDir is explicitly the default if AllPtrFieldsNil(VS) - that is intended.  Ahh, but that is patched up in validation, which means this whole check can probably be removed!  Yay.

File: pkg/volume/git_repo/git_repo.go (hunk 4 above)
Comment (thockin) [inline pkg/volume/git_repo/git_repo.go]: OK, fair enough.  I don't get the gofmt heuristic, then :)

File: pkg/volume/host_path/host_path.go (hunk 5 above)
Comment (thockin) [inline pkg/volume/host_path/host_path.go]: OK

File: pkg/volume/iscsi/iscsi.go (hunk 6 above)
Comment (thockin) [inline pkg/volume/iscsi/iscsi.go]: Sure, I just thought that had been done.  NP

File: pkg/volume/nfs/nfs.go (hunk 7 above)
Comment (thockin) [inline pkg/volume/nfs/nfs.go]: OK

File: pkg/volume/plugins.go (hunk 8 above)
Comment (thockin) [inline pkg/volume/plugins.go]: As per the above, I guess making them pointers is not required.

File: pkg/volume/plugins_test.go (hunk 11 above)
Comment (thockin) [inline pkg/volume/plugins_test.go]: my screen has way more pixels across than it does up-down.  :)

**Comment by thockin** [comment]
Ping this PR when you're ready for round 2.

File: pkg/volume/empty_dir/empty_dir.go (hunk 2 above)
Comment (markturansky) [inline pkg/volume/empty_dir/empty_dir.go]: Cool, I verified defaults.go and then removed this check entirely.

**Comment by markturansky** [comment]
Feedback addressed.  Round Two may commence.

File: pkg/volume/plugins.go (hunk 8 above)
Comment (markturansky) [inline pkg/volume/plugins.go]: The pointers complicated the CanSupport methods in the plugins, actually, because they'd have to check first for nil VS/PSV and then the nil src.

Comment (thockin) [inline pkg/volume/plugins.go]: Still suggest a rename.  `type Spec` or `type APISpec` or `type FromAPI` or something

Comment (markturansky) [inline pkg/volume/plugins.go]: yes, missed that on the 2nd commit, but refresh and take a look at the 3rd.

----- Original Message -----

**Comment by thockin** [comment]
LGTM. Squash please.

**Comment by markturansky** [comment]
Rebased and squashed.  Waiting for green.

**Comment by markturansky** [comment]
@thockin green. punch it.
